Fix warning and bug with update with reload=False #9

sserov opened the pull request:
Massively refactored the code into some classes, added sleep to deal with too frequent requests and fixed a bug with update with reload=False. Also fixed warnings in calculate_statistics for 1 week with quantiles (for 1 day may emerge because the market was closed the previous day)

sserov committed:
Refactored the functions from analysis.py to the History class

sserov committed:
Massively refactored the code, created MarketWrapper class and divided the logic

sserov committed:
Included sleep 5 sec in case the server tells there are too many requests + small fixes #9

sserov committed:
Fixed a bug in analysis.py with incorrect update with reload=False #9

Alexander-Serov commented:
> Massively refactored the code into some classes, added sleep to deal with too frequent requests and fixed a bug with update with reload=False. Also fixed warnings in calculate_statistics for 1 week with quantiles (for 1 day may emerge because the market was closed the previous day)

Can be more than 1 day if a longer weekend?

Alexander-Serov commented on history.py:
```diff
@@ -17,342 +17,21 @@
-import inspect
 import math
 import os
-import pathlib
```
В следующий раз постарайся перенос из файла в файл и изменение кода поместить в разные коммиты. Иначе трудно отличить перенос от модификации.

sserov replied:
Окей

Alexander-Serov commented on history.py:
```diff
@@ -457,6 +125,216 @@ def _save_data(self):
+                slept = False
+            except Exception as e:
+                print(e)
+                if figi not in utils.OBSOLETE_TICKERS.values():
```
Может быть для устаревших не стоит и пытаться получить свечи?

sserov replied:
Да, думаю, смысла нет, потому что портфель и можно оценивать, но купить все равно больше нельзя

sserov replied:
Пуфи выяснил, что они и от сервера-то в списке етфов больше не приходят. Но Пуфи на всякий случай перенес эту проверку в получение списка

sserov replied:
Починено в 15c3fa4

Alexander-Serov commented on history.py:
```diff
@@ -457,6 +125,216 @@ def _save_data(self):
+                )
+
+                if slept:
+                    print("LOADED AFTER SLEEP")
```
Ты можешь поместить это в лог, но вряд ли нужен принт.

sserov replied:
Убрано в 15c3fa4

Alexander-Serov commented on history.py:
```diff
@@ -457,6 +125,216 @@ def _save_data(self):
+                if figi not in utils.OBSOLETE_TICKERS.values():
+                    utils.log_to_file(f"Unable to load history for figi={figi}")
+                    utils.log_to_file(e)
+                print("Sleep 5 seconds")
```
И это тоже можно спрятать, нет?

sserov replied:
Убрано в 15c3fa4

Alexander-Serov commented on history.py:
```diff
@@ -457,6 +125,216 @@ def _save_data(self):
+                    utils.log_to_file(f"Unable to load history for figi={figi}")
+                    utils.log_to_file(e)
+                print("Sleep 5 seconds")
+                time.sleep(5)
```
Давай вытащим константу на верх файла?

sserov replied:
Вытащено в c056176

Alexander-Serov commented on history.py:
```diff
@@ -457,6 +125,216 @@ def _save_data(self):
+        figi = self.market_wrapper.get_figi_for_ticker(ticker)
+        return self.get_figi_history(figi, start, end, interval)
+
+    def get_figi_history(self, figi, start, end, interval):
```
```suggestion
    def get_figi_history(self, figi:str, start:datetime, end:datetime, interval: str):
```

Alexander-Serov commented on history.py:
```diff
@@ -457,6 +125,216 @@ def _save_data(self):
+        list
+            List of ETF tickers.
+        """
+        # print('C1', start, end)
```
```suggestion
```

sserov replied:
890b994

Alexander-Serov commented on history.py:
```diff
@@ -457,6 +125,216 @@ def _save_data(self):
+            ticker = etf.ticker
+            tickers.append(ticker)
+
+            # print(etf, figi, ticker)
```
```suggestion
```

sserov replied:
890b994

Alexander-Serov commented on history.py:
```diff
@@ -457,6 +125,216 @@ def _save_data(self):
+                    f"Getting ETF history for figi={figi} from {start} till {end},"
+                    f" interval={freq}."
+                )
+            # print(i, end=" ")
```
```suggestion
```

sserov replied:
890b994

sserov commented:
> 
> 
> > Massively refactored the code into some classes, added sleep to deal with too frequent requests and fixed a bug with update with reload=False. Also fixed warnings in calculate_statistics for 1 week with quantiles (for 1 day may emerge because the market was closed the previous day)
> 
> Can be more than 1 day if a longer weekend?

No, the problem was that in these cases the 1.0 quantile was used instead of 0.99 in longer periods. The warning wouldn't emerge if there was at least one trading day during a period

Alexander-Serov commented on history.py:
```diff
@@ -457,6 +125,216 @@ def _save_data(self):
+                figi=figi, start=start, end=end, interval=freq
+            )
+
+            if one_etf_history.empty:
```
Добавим комментарий, в каком случае это происходит?

sserov replied:
a1c660c

Alexander-Serov commented on history.py:
```diff
@@ -457,6 +125,216 @@ def _save_data(self):
+            if all_etfs_history.empty:
+                all_etfs_history = one_etf_history
+            else:
+                # all_etfs_history = all_etfs_history.merge(one_etf_history,
```
можно убрать

sserov replied:
Убрано

Alexander-Serov commented on history.py:
```diff
@@ -457,6 +125,216 @@ def _save_data(self):
+        )
+
+        # For daily forecasts drop hours
+        # end = end.replace(hour=0, minute=0, second=0,microsecond=0)
```
Почему это неактивно?

sserov replied:
Твой старый комментарий. Я только рефакторил и не откомменчивал ничего нового

sserov replied:
Понятия не имею, почему неактивно

sserov replied:
7e7f9b0 Удалено

Alexander-Serov commented on history.py:
```diff
@@ -457,6 +125,216 @@ def _save_data(self):
+        #                   RuntimeWarning)
+
+        # Drop all _data points after the end date (because the API does not return exactly
+        # what was requested)
```
Непонятно почему здесь неактивен этот код. Если больше не нужен, надо убрать. Если нужен, то вернуть.

sserov replied:
Твой коммент, не откомменчивал

sserov replied:
Убрано

sserov replied:
7e7f9b0e7e010b73da6ec93ebf4a8e8812e08e08

Alexander-Serov commented on history.py:
```diff
@@ -466,20 +344,22 @@ def update(self, reload=False):
         # date in the database
         if self._data.empty or reload:
-            start_date = EARLIEST_DATE
+            start_date = utils.EARLIEST_DATE
```
Мне кажется константы можно импортировать без родителя. Москоу таймзоун все равно одна на всех.

sserov replied:
a1c660c

Alexander-Serov commented on history.py:
```diff
@@ -503,7 +383,7 @@ def update(self, reload=False):
             # Convert the time column to time because after merge it changes
             # to object
-            merged.time = pd.to_datetime(self._data.time)
+            merged.time = pd.to_datetime(merged.time)
```
Неплохо!

Alexander-Serov commented on history.py:
```diff
@@ -557,39 +440,39 @@ def calculate_statistics(self, position="c"):
             "max_1w": self._data.loc[self._data.time >= filter_1w, ["figi", "h"]]
             .groupby(by="figi")
-            .quantile(1)["h"]
+            .quantile(max_quantile)["h"]
```
ну здесь спокойно можно было оставить 0 и 1, нет?

sserov replied:
Из-за этого возникали наны, вроде бы. Искал этот баг. Правда, возможно, он был наведенный от какой-то проблемы. Кажется, это что-то решало. Если критично, вернем

sserov replied:
Вернул be47918

Alexander-Serov commented on history.py:
```diff
@@ -557,39 +440,39 @@ def calculate_statistics(self, position="c"):
             .first()[position]
             .to_dict(),
-            "1d": self._data.loc[self._data.time <= filter_1d, ["figi", position]]
+            "1d": self._data.loc[self._data.time >= filter_1d, ["figi", position]]
```
На этом гитхабе тупо, что я ставлю согласиться на изменения, но если потом отправляю ревью, они сбрасываются. Их типа надо применить перед комменатриями.

sserov replied:
890b994 Вернуть <=

Alexander-Serov commented on history.py:
```diff
@@ -557,39 +440,39 @@ def calculate_statistics(self, position="c"):
-        #                       by='figi'))
-        # print(self._data[self._data.ticker == 'FXUS'].time.unique())
 
+        # print(self._data[self._data.time >= filter_1w].groupby(by='figi'))
```
Нужно убрать если не используется.

sserov replied:
Убрано 890b994

Alexander-Serov commented on history.py:
```diff
@@ -557,39 +440,39 @@ def calculate_statistics(self, position="c"):
                 warnings.warn(
-                    f"Not all figis were found for the column `"
-                    f"{key}`. The analysis may be incorrect."
+                    f"Not all figis were found for the column '{key}'."
```
ну уж

sserov replied:
Пофикшено f1a5ee5

Alexander-Serov commented on history.py:
```diff
@@ -714,3 +597,24 @@ def recommend_other(self, update: bool = True, _print=False, reload=False):
+        else:
+            last_date = None
+
+        return last_date
```
```suggestion
        return self._data.time.max() if not self._data.empty else None
```

Alexander-Serov commented on main-analysis.ipynb:
```diff
@@ -71,9 +38,9 @@
    "source": [
-    "plt.figure()\n",
-    "plt.plot(ticker_history[\"time\"].values, ticker_history[\"c\"])\n",
-    "plt.title(ticker)"
```
Больше не плотим?

sserov replied:
Ну там чуть-чуть ячейки поменял. Этот ноутбук идет прицепом пока что

Alexander-Serov commented on main-analysis.ipynb:
```diff
@@ -98,10 +77,9 @@
-    "# display(hist_daily.data.time.dtype)\n",
-    "hist_daily.update(reload=0)\n",
-    "hist_daily.data"
+    "plt.figure()\n",
```
А, типа тут. ок.

Alexander-Serov commented on marketWrapper.py:
```diff
@@ -0,0 +1,96 @@
+from functools import lru_cache
```
Пьер бы тут написал, что имя файла должно быть snake_case, а имя класса - CamelCase.

sserov replied:
Окей, ты поменял cc2f65e

Alexander-Serov commented on marketWrapper.py:
```diff
@@ -0,0 +1,96 @@
+            figi = res[0].figi
+        else:
+            figi = None
+        return figi
```
```suggestion
        return res[0].figi if res else None
```

Alexander-Serov commented on marketWrapper.py:
```diff
@@ -0,0 +1,96 @@
+            if ticker is None:
+                raise ValueError("Either ticker or figi should be provided.")
+            else:
+                figi = self.get_figi_for_ticker(ticker)
```
pylint не сказал здесь что не нужен элс после райз??

sserov replied:
Убрано в cc2f65e

Alexander-Serov commented on marketWrapper.py:
```diff
@@ -0,0 +1,96 @@
+            current_price = payload.close_price
+        else:
+            order_response = payload.asks[0]
+            # print('debug-1', get_ticker_for_figi(figi), order_response)
```
```suggestion
```

Alexander-Serov commented on marketWrapper.py:
```diff
@@ -0,0 +1,96 @@
+            # print('debug-1', get_ticker_for_figi(figi), order_response)
+            current_price = order_response.price
+
+            # TODO not tested
```
?

sserov replied:
Если ты про тудушку, то это тоже твой старый коммент к этой функции

sserov replied:
cc2f65e

Alexander-Serov commented on utils.py:
```diff
@@ -0,0 +1,47 @@
+LOCAL_TIMEZONE = dt.datetime.now(dt.timezone.utc).astimezone().tzinfo
+MOSCOW_TIMEZONE = pytz.timezone("Europe/Moscow")
+EARLIEST_DATE = dt.datetime.fromisoformat("2013-01-01").replace(tzinfo=MOSCOW_TIMEZONE)
+# TODO: In the received results, Moscow timezone sometimes appears as +2:30 and
```
Это новый комментарий? Это очень странно что разница в полчаса.

sserov replied:
Это старый твой коммент. Не проверял это, но с виду выглядело, что всё нормально с данными в этом плане

Alexander-Serov replied:
Ну тогда предлагаю выкинуть.

sserov replied:
f1a5ee5

Alexander-Serov commented:
В целом, я так понял это приведения в порядок. Трудно было проверить какие-то изменения по сути, если они были, так как всё вместе. В целом, выглядит нормально, хотя, может, некоторые пояснения о том, как каждая структура должна использоваться, не помешали бы.

sserov committed:
Moved check for obsolete tickers to the _get_all_etfs() function and removed redundant checks for obsoleteness

sserov committed:
Remove comment from marketWrapper.py

Co-authored-by: Alexander Serov <asserov@gmail.com>

sserov committed:
Fixed a bug with merge data with different timezones in column "time"

sserov committed:
Moved sleep constant to utils.py

sserov committed:
Update return in get_figi_for_ticker() function

Co-authored-by: Alexander Serov <asserov@gmail.com>

Alexander-Serov commented on history.py:
```diff
@@ -385,6 +385,9 @@ def update(self, reload=False):
 
+            new_data.loc[:, "time"] = new_data["time"].dt.tz_convert(
+                old_data["time"].dt.tz
+            )  # To UTC
```
Хорошо.
Но во-первых слева можно без тайм. А во-вторых, если ту утс, то так и конверть напрямую в `pytz.utc`?

sserov replied:
В смысле без .loc? Окей, можно убрать. Пуфи там попробовал dt.tz_convert(None) (написано, что None -- это UTC) -- не взлетело. А так работает

sserov replied:
все починил, спасибо Нафи.

Alexander-Serov committed:
Rename MarketWrapper and add a set of tests.

Alexander-Serov committed:
Small

sserov committed:
Fixed incorrect fix <= for >=. Removed comments. Added sleep timer with count of restarts to market get functions

sserov committed:
Remove comments and write warning in one line

sserov committed:
Update get_figi_history definition

Co-authored-by: Alexander Serov <asserov@gmail.com>

sserov committed:
Update return in last_date property

Co-authored-by: Alexander Serov <asserov@gmail.com>

sserov committed:
Removed redundant variable success, added comment, imported constants without the file name

sserov committed:
Remove slept variable

sserov committed:
Revert 0 and 1 quantiles and added if for payload with empty candles

sserov committed:
Deleted Nafi's inactive code

sserov committed:
Removed slept variable

sserov committed:
Fixed identation bug from rebase and added UTC conversion for the new data if update=True

Alexander-Serov commented on utils.py:
```diff
@@ -0,0 +1,48 @@
+colorama.init()
+
+# Load the token
+MAIN_FOLDER = pathlib.Path(os.path.dirname(os.path.realpath(__file__)))
```
Пьер ровно это исправил Луи два дня назад. Слово в слово ))

sserov replied:
Отлично

Alexander-Serov commented on main-analysis.ipynb:
```diff
@@ -71,9 +38,9 @@
-    "plt.title(ticker)"
+    "hist_daily = History(interval=\"day\")\n",
+    "hist_daily.update(reload=True)\n",
+    "hist_daily.data  # Should be 28600 + rows"
```
Я понял, что это значит, но не уверен, зачем нам оставлять здесь этот комменатарий?

sserov replied:
Это было на случай, если падала загрузка по количеству запросов. Убрано в 1bad1cbb

Alexander-Serov committed:
Rename MarketWrapper to Marker.
Other small changes.

Alexander-Serov reviewed (approved):
В целом всё хорошо. Проверь мои несколько последних коммитов и можно мержить.

sserov committed:
Removed comment in notebook and renamed _print parameter of recommend_simple()

sserov committed:
Added tests for get_all_etfs() and test_get_candles() functions of Market class

Alexander-Serov commented on tests/test_market.py:
```diff
@@ -0,0 +1,118 @@
+        assert etf.type == "Etf"
+        assert etf.ticker not in OBSOLETE_TICKERS
+        assert 0 < etf.min_price_increment < np.inf
+        assert isinstance(etf.lot, int) and 0 < etf.lot < np.inf
```
Что такое лот?

sserov replied:
Судя по всему, это минимальное количество паёв, которое можно купить / продать за сделку. Собственно, лот

Alexander-Serov commented on tests/test_market.py:
```diff
@@ -0,0 +1,118 @@
+        for st_end_int in start_end_interval:
+            candles = market.get_candles(
+                figi,
+                _from=st_end_int["start"].isoformat(),
```
Жалко что оно принимает строки, а не дэйт-таймы.

sserov replied:
Что есть, то есть

Alexander-Serov replied:
Ну это же наша функция, нет?

Alexander-Serov commented on tests/test_market.py:
```diff
@@ -0,0 +1,118 @@
+                interval=st_end_int["interval"],
+            ).payload.candles
+
+            assert isinstance(candles, list)
```
and candles ?

Alexander-Serov replied:
Или если рынок закрыт, то пустые и ничего не надо тестить?

sserov replied:
Да, он возвращает пустой список, если рынок закрыт и был закрыт весь запрошенный период

Alexander-Serov committed:
Small changes in the new tests.

Alexander-Serov reviewed (commented):
Хорошие тесты, мне понравились.

Alexander-Serov reviewed (approved):
Я думаю можно мержить (я бы только наверное сквошнул).

sserov commented:
> 
> 
> Я думаю можно мержить (я бы только наверное сквошнул).

А потом не случится как тогда, что нужно будет всё ребэйзить 3 часа?

sserov commented:
> 
> 
> Хорошие тесты, мне понравились.

Отлично

sserov commented:
Есть кнопка сквош энд мёрдж

sserov commented:
Тыкаю типа?

Alexander-Serov commented:
> > 
> 
> > 
> 
> > Я думаю можно мержить (я бы только наверное сквошнул).
> 
> 
> 
> А потом не случится как тогда, что нужно будет всё ребэйзить 3 часа?

Ну в целом это часть жизни. 
Если ты ничего не менял в этой части, то и ребэйз проходит молча.

Alexander-Serov commented:
> Тыкаю типа?

Думаю, можно.
